alf/examples/grid_search.py

```diff
@@ -0,0 +1,101 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import multiprocessing
+import itertools
+import json
+
+from absl import app
+from absl import flags
+from absl import logging
+import gin
+from alf.examples.main import train_eval
+from alf.utils import common
+
+flags.DEFINE_string('search_file', None,
+                    'Path to the grid search config file.')
+FLAGS = flags.FLAGS
+r"""Grid search.
+
+To run grid search on ddpg for gym Pendulum:
+```bash
+python grid_search.py \
+  --root_dir=~/tmp/ddpg_pendulum \
+  --search_file=ddpg_grid_search.json \
+  --gin_file=ddpg_pendulum.gin \
+  --gin_param='create_environment.num_parallel_environments=8' \
+  --alsologtostderr
+```
+"""
+
+
+class GridSearch(object):
+    """Grid Search"""
+
+    def __init__(self, conf_file):
+        """
+        Args:
+            conf_file (str): Path to the config file. Config file must be json format,
+                and `parameters` is required. Parameters consist of key and iterable values.
+                see `ddpg_grid_search.json` for a good example
+                for more details.
+        """
+        self._conf_file = conf_file
+
+    @staticmethod
+    def parse_config(conf_file):
+        with open(conf_file) as f:
+            keys = []
+            values = []
+            conf = json.loads(f.read())
+            for key, value in conf['parameters'].items():
+                if isinstance(value, str):
+                    value = eval(value)
+                keys.append(key)
+                values.append(value)
+        return keys, values
+
+    def run(self):
+        processes = []
+        para_keys, para_values = GridSearch.parse_config(self._conf_file)
+        for values in itertools.product(*para_values):
```

Comment:
should have an option to limit the number of running processes

Reply:
add `max_worker_num` to limit the number of running processed